sampo/userinput/parser/history.py

```diff
@@ -186,56 +193,93 @@ def get_all_seq_statistic(history_data: pd.DataFrame, tasks_structure_df: pd.Dat
 
-    predecessors_ids_lst, predecessors_types_lst, predecessors_lags_lst = [], [], []
+    predecessors_ids_lst, predecessors_types_lst, predecessors_lags_lst, predecessors_counts_lst = [], [], [], []
 
     for task_id, pred_info_lst in connections_dict.items():
-        pred_ids_lst, pred_types_lst, pred_lags_lst = zip(*pred_info_lst)
+        if len(pred_info_lst) > 0:
+            pred_ids_lst, pred_types_lst, pred_lags_lst, pred_counts_lst = map(list, zip(*pred_info_lst))
+        else:
+            pred_ids_lst, pred_types_lst, pred_lags_lst, pred_counts_lst = ['-1'], ['-1'], ['-1'], ['-1']
+
+        if str(task_id) in tasks_df.index:
+            if tasks_df.loc[str(task_id), 'predecessor_ids'] != ['-1']:
+                if expert_connections_info:
+                    predecessors_ids_lst.append(tasks_df.loc[str(task_id), 'predecessor_ids'])
+                    predecessors_types_lst.append(tasks_df.loc[str(task_id), 'connection_types'])
+                    predecessors_lags_lst.append(tasks_df.loc[str(task_id), 'lags'])
+                    predecessors_counts_lst.append(pred_counts_lst)
+                    continue
+                if change_connections_info:
+                    predecessors_ids_lst.append(tasks_df.loc[str(task_id), 'predecessor_ids'])
+                else:
+                    predecessors_ids_lst.append(pred_ids_lst)
+            else:
+                predecessors_ids_lst.append(pred_ids_lst)
+        else:
+            predecessors_ids_lst.append(pred_ids_lst)
```

Comment:
This relational `else` statements can be rewritten to one statement after the outer `if`

Reply:
Nah. That would defeat all logic.